### PR TITLE
fix(options): deprecate `nvim[_buf|_win]_[gs]et_option`

### PR DESCRIPTION
**Pull request by lewis6991**
Replaced with `nvim_[gs]et_option_value`.

### CODE REVIEW

**Comment by zeertzjq** [comment]
IIRC `nvim_win_set_option` cannot be replaced with `nvim_set_option_value`. This has been mentioned in some issue.

**Comment by zeertzjq** [comment]
It's #20045

**Comment by lewis6991** [comment]
That issue was wrong, see my comments. `nvim_win_set_option` sets the non-local option so is replaceable with `nvim_set_option_value` without `scope` being provided. `scope` is needed to match the behaviour of `setlocal`.

**Review by clason — APPROVED** [review]

**Review by clason — CHANGES_REQUESTED** [review]
Thinking about this, we should make sure in this PR already that the deprecated functions aren't used in our codebase any more. (A sample ripgrep showed some tests still using them.)

**Comment by lewis6991** [comment]
I changed everything in runtime. Didn't check the tests though.

**Comment by clason** [comment]
Yes, I saw that, and I don't think you missed any in runtime; I was specifically referring to the tests.

**Comment by lewis6991** [comment]
Ok I'll do the tests.

**Comment by lewis6991** [comment]
That was hard. Automated a fair bit of it. And I think there's actually a bug in the new functions.

File: test/functional/ui/messages_spec.lua [hunk 1]
```diff
@@ -1013,9 +1013,9 @@ describe('ui/builtin messages', function()
     -- ignore final whitespace inside string
     -- luacheck: push ignore
     eq([[--- Syntax items ---
-vimComment     xxx match /\s"[^\-:.%#=*].*$/ms=s+1,lc=1  excludenl contains=@vimCommentGroup,vimCommentString 
-                   match /\<endif\s\+".*$/ms=s+5,lc=5  contains=@vimCommentGroup,vimCommentString 
-                   match /\<else\s\+".*$/ms=s+4,lc=4  contains=@vimCommentGroup,vimCommentString 
+vimComment     xxx match /\s"[^\-:.%#=*].*$/ms=s+1,lc=1  excludenl contains=@vimCommentGroup,vimCommentString
+                   match /\<endif\s\+".*$/ms=s+5,lc=5  contains=@vimCommentGroup,vimCommentString
+                   match /\<else\s\+".*$/ms=s+4,lc=4  contains=@vimCommentGroup,vimCommentString
```
Comment (clason) [inline test/functional/ui/messages_spec.lua:1018]: I think that final whitespace is intentional...

Comment (lewis6991) [reply test/functional/ui/messages_spec.lua:1018]: Thanks. My setup probably stripped the lines under my feet.

File: test/functional/ui/screen_basic_spec.lua [hunk 2]
```diff
@@ -157,18 +157,18 @@ local function screen_tests(linegrid)
...
+      it('an RPC call to nvim_set_option_value in a hidden buffer', function()
+        meths.set_option_value('autoindent', true, {buf = buf2})
         command('redraw!')
         screen:expect(function()
           eq(expected, screen.title)
         end)
       end)
 
-      it('a Lua callback calling nvim_buf_set_option in a hidden buffer', function()
+      it('a Lua callback calling nvim_buf_set_option_value in a hidden buffer', function()
```
Comment (zeertzjq) [inline test/functional/ui/screen_basic_spec.lua:168]:
```suggestion
      it('a Lua callback calling nvim_set_option_value in a hidden buffer', function()
```

File: test/functional/ui/float_spec.lua [hunk 3]
```diff
@@ -126,7 +126,7 @@ describe('float window', function()
 
   it('opened with correct width', function()
     local width = exec_lua([[
-      vim.api.nvim_set_option("winwidth", 20)
+      vim.go.winheight = 20
```
Comment (zeertzjq) [inline test/functional/ui/float_spec.lua:129]:
```suggestion
      vim.go.winwidth = 20
```

File: test/functional/terminal/tui_spec.lua [hunk 4]
```diff
@@ -1403,9 +1403,9 @@ describe('TUI', function()
       {3:-- TERMINAL --}                                    |
     ]]
     screen:expect(singlewidth_screen, attrs)
-    child_session:request('nvim_set_option', 'ambiwidth', 'double')
+    child_session:request('nvim_set_option_value', 'ambiwidth', 'double', {scope='global'})
     screen:expect(doublewidth_screen, attrs)
-    child_session:request('nvim_set_option', 'ambiwidth', 'single')
+    child_session:request('nvim_set_option_value', 'ambiwidth', 'single', {scope='global'})
```
Comment (zeertzjq) [inline test/functional/terminal/tui_spec.lua:1408]: `'ambiwidth'` is a global option so `scope='global'` is not needed.

File: test/functional/terminal/scrollback_spec.lua [hunk 5]
```diff
@@ -383,8 +383,8 @@ describe("'scrollback' option", function()
 
   local function set_fake_shell()
     -- shell-test.c is a fake shell that prints its arguments and exits.
-    nvim('set_option', 'shell', testprg('shell-test'))
-    nvim('set_option', 'shellcmdflag', 'EXE')
+    nvim('set_option_value', 'shell', testprg('shell-test'), {scope='global'})
+    nvim('set_option_value', 'shellcmdflag', 'EXE', {scope='global'})
```
Comment (zeertzjq) [inline test/functional/terminal/scrollback_spec.lua:387]: ditto

**Comment by lewis6991** [comment]
`nvim_get_option_value` has some subtle differences to `nvim_[buf_|win_]get_option` due to it having a very different implementation. I'm currently working on fixing this so it uses the same implementation. So this might take a bit more work but it should bring some perf improvements for free.

File: src/nvim/api/options.c [hunk 6]
```diff
@@ -254,17 +222,19 @@ void nvim_set_option(uint64_t channel_id, String name, Object value, Error *err)
 
 /// Gets the global value of an option.
 ///
+/// @deprecated
 /// @param name     Option name
 /// @param[out] err Error details, if any
 /// @return         Option value (global)
 Object nvim_get_option(String name, Arena *arena, Error *err)
```
Comment (justinmk) [inline src/nvim/api/options.c:229]: if possible, ideally this would live in `api/deprecated.c`

**Comment by ddickstein** [comment]
> That issue was wrong, see my comments. nvim_win_set_option sets the non-local option so is replaceable with nvim_set_option_value without scope being provided. scope is needed to match the behaviour of setlocal.

@lewis6991 Just to clarify, https://github.com/neovim/neovim/issues/20045 says at the top that it was written for Neovim 0.7.0, whose version of `nvim_set_option_value` did not support `win` or `buf`, so it could not replace `nvim_[buf|win]_set_option`. That issue is correct for the version of Neovim for which it was written.

**Comment by lewis6991** [comment]
But the issue was also raised after v0.8 had been released. So the fact it was for v0.7 was irrelevant from a maintainers POV. The state of master is the only thing that is relevant for this issue tracker as we don't maintain old releases.

**Pull request closed without merging**